### ISSUE TITLE
Broken jsx element with not static type with ts 3.2.1

### ISSUE THREAD

**Issue by dfilatov** [Microsoft/TypeScript#28768]
**TypeScript Version:**  3.2.1

**Search Terms:**
jsx

**Code**

```tsx
import * as React from 'react';

function render(url?: string): React.ReactNode {
    const Tag = url? 'a' : 'button';
    return <Tag>test</Tag>;
}
```

**Expected behavior:**
Everything is ok. TypeScript 3.1.6 compiles it properly.

**Actual behavior:**
error: JSX element type 'Tag' does not have any construct or call signatures. [2604]

**Comment by devrelm** [comment]
Related: #28631

**Comment by weswigham** [comment]
It's because we started strongly checking tag calls such as these, and the `'a`` tag's properties and `'button'` tag's properties are heavily dissimilar - they're all optional, so a call like this can still succeed at runtime, but in the typesystem at present the two signatures are irreconcilable. Needs a full fix of #7294 to work well.

**Comment by gsbelarus** [comment]
Me too! We have switched back to TS 3.1.6 because of code like below stopped compiling:

```
class LinkCommandBarButton extends BaseComponent<ILinkCommandBarButtonProps> {
  public render(): JSX.Element {
    const { defaultRender: DefaultRender = CommandBarButton, link, supText, ...buttonProps } = this.props;

    const onRenderText = supText ? (props: IButtonProps) => <>{props.text}<sup>{supText}</sup></> : undefined;

    return (
      <Link to={link}>
        <DefaultRender {...buttonProps} onRenderText={onRenderText} />
      </Link>
    );
  }
};
```

**Comment by rhys-vdw** [comment]
Please consider this now broken example:

```ts
import React, {
  ReactType,
  StatelessComponent,
} from "react"
import classNames from "classnames"

interface Props {
  className?: string
}

export function createStyledComponent<P extends { className: string }>(
  Component: ReactType<P>,
  className: string
): StatelessComponent<Props> {
  return props => (
    <Component className={classNames(className, props.className)}>
      {props.children}
    </Component>
  )
}
```

```
ERROR in [at-loader] ./app/assets/javascripts/utilities/component-factory.tsx:19:6
    TS2604: JSX element type 'Component' does not have any construct or call signatures.
```

Is this something that should be addressed in @types/react?

**Comment by Kovensky** [comment]
Nothing that can be done on the `@types/react` side, or rather, what can be done is already done -- `ReactType<P>` will work correctly with `React.createElement`.

![image](https://user-images.githubusercontent.com/73085/49348446-da4b1780-f6e8-11e8-9026-cc3665d3c219.png)

Note that your `P extends { className: string }` will actually reject all DOM elements because their `className` includes `undefined` which is not assignable to `string`.

![image](https://user-images.githubusercontent.com/73085/49348420-a5d75b80-f6e8-11e8-925c-f6fb757c68c9.png)

**Comment by RyanCavanaugh** [comment]
@weswigham can we get a fresh read on this?

**Comment by weswigham** [comment]
The OP's example should have been fixed by #29011 (and testing shows it is). @rhys-vdw's example [here](https://github.com/Microsoft/TypeScript/issues/28768#issuecomment-443569007) also works. @azizhk's example probably typechecks better than he wants, because we correctly reject the spread of a `React.InputHTMLAttributes<HTMLTextAreaElement>` as `A`'s arguments, because a `"input"` element won't be happy getting a `"textarea"`'s `onChange` handler.

**Comment by azizhk** [comment]
Hi, I just upgraded to 
```
"@types/react": "^16.8.7",
"@types/react-dom": "^16.8.2",
"typescript": "^3.3.3333",
```

I can use this for string components now `const A = condition ? "input" : "textarea"`,

But I cannot use this for components which have generics

```tsx
class OptionGroupCheckBox<OptionType> extends React.PureComponent<OptionGroupProps<OptionType>> {
  // methods
}

class OptionGroupRadio<OptionType> extends React.PureComponent<OptionGroupProps<OptionType>> {
  // methods
}

const OptionGroup = condition ? OptionGroupCheckBox : OptionGroupRadio;
const vdom = (
  <OptionGroup
    // OptionGroupProps<OptionType>
  />
)

```
